Various bug fixes relating to shutdown handling

Frogging101 opened the pull request:
I have created patches to fix a few issues that can occur when shutting down rsyslog. These issues include:
- A race condition that can cause rsyslogd to block for the full janitor interval (10 minutes) before responding to a signal such as SIGTERM or SIGHUP.
- A race condition that can cause tcpsrv to segfault when shutting down
- Cleaning up tcpsrv properly if its thread is cancelled

See the commit messages for more details. Each commit represents a logical code change, with a detailed explanation in the commit message.

----
I created these changes in the course of my employment at Ciena Corporation (https://www.ciena.com/), and am submitting them to the rsyslog project with permission and on behalf of Ciena. These changes are provided to rsyslog under the terms of the GNU General Public License (Version 3).

I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

rgerhards commented:
Thanks for the pull request! Would it be possible that you release the patches under ASL 2.0? We are currently trying to move towards that license and adding more GPL-only code adds stumbling stones to that process (and we would need to remove that code in the longer term in any case).

side-note: I have deliberately not yet taken a look at the patches as I do not want to accidently pollute my mind with the code, in case we need to take the bug descriptions and find our own fixes. Thus no comments on CI.

Frogging101 commented:
> Thanks for the pull request! Would it be possible that you release the patches under ASL 2.0? We are currently trying to move towards that license and adding more GPL-only code adds stumbling stones to that process (and we would need to remove that code in the longer term in any case).
> 
> side-note: I have deliberately not yet taken a look at the patches as I do not want to accidently pollute my mind with the code, in case we need to take the bug descriptions and find our own fixes. Thus no comments on CI.

I've only modified these files:
- `runtime/tcpsrv.c`
- `tools/rsyslogd.c`

Are you sure you need the changes under ASL?

davidelang commented:
On Thu, 24 Oct 2019, John Brooks wrote:

>> Thanks for the pull request! Would it be possible that you release the patches under ASL 2.0? We are currently trying to move towards that license and adding more GPL-only code adds stumbling stones to that process (and we would need to remove that code in the longer term in any case).
>>
>> side-note: I have deliberately not yet taken a look at the patches as I do not want to accidently pollute my mind with the code, in case we need to take the bug descriptions and find our own fixes. Thus no comments on CI.
>
> I've only modified these files:
> - `runtime/tcpsrv.c`
> - `tools/rsyslogd.c`
>
> Are you sure you need the changes under ASL?

Yes please, any significant work needs to be under a compatible license.

David Lang

Frogging101 commented:
As you wish. I will ask our legal team for approval to provide the changes under the Apache Software License 2.0.

Frogging101 commented:
These changes are now provided to rsyslog under the terms of the Apache Software License 2.0.

Frogging101 commented:
I see that build failure with GCC 8.

```
In file included from tcpsrv.c:60:
tcpsrv.c: In function 'processWorkset':
rsyslog.h:625:26: error: variable 'iRet' might be clobbered by 'longjmp' or 'vfork' [-Werror=clobbered]
 #define DEFiRet rsRetVal iRet = RS_RET_OK
                          ^~~~
tcpsrv.c:733:2: note: in expansion of macro 'DEFiRet'
  DEFiRet;
  ^~~~~~~
tcpsrv.c: In function 'Run':
rsyslog.h:625:26: error: variable 'iRet' might be clobbered by 'longjmp' or 'vfork' [-Werror=clobbered]
 #define DEFiRet rsRetVal iRet = RS_RET_OK
                          ^~~~
tcpsrv.c:910:2: note: in expansion of macro 'DEFiRet'
  DEFiRet;
  ^~~~~~~
cc1: all warnings being treated as errors
```

The warning could be spurious; there are a number of bugs filed about `-Wclobbered`: https://gcc.gnu.org/bugzilla//show_bug.cgi?id=21161 (see related bugs under "See Also").

It may be possible to rearrange the code so the compiler doesn't complain. Not sure how yet, though.

rgerhards commented:
> It may be possible to rearrange the code so the compiler doesn't complain. Not sure how yet, though.

I had a quick look. This seems to be caused by the POSIX threads cancel handler macros. That's typically the case. They involve longjump and so the warning is valid. So far this was solved by writing a function that contains only the cancel handler macros plus a call to the real function.

rgerhards commented:
@Frogging101 any interest in pursuing this further?

Frogging101 commented:
For sure. It's on my to-do list :) 

On November 14, 2019 8:49:45 AM EST, Rainer Gerhards <notifications@github.com> wrote:
>@Frogging101 any interest in pursuing this further?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/rsyslog/rsyslog/pull/3924#issuecomment-553895788

-- 
Sent from my BlackBerry PRIV with K-9 Mail

Frogging101 commented:
I think I've fixed it now.

Frogging101 commented:
This is what I changed:
```diff
diff --git a/runtime/tcpsrv.c b/runtime/tcpsrv.c
index 1b76e1492..f761ff7cc 100644
--- a/runtime/tcpsrv.c
+++ b/runtime/tcpsrv.c
@@ -719,6 +719,20 @@ wrkr(void *const myself)
        return NULL;
 }
 
+/* This has been factored out from processWorkset() because
+ * pthread_cleanup_push() invokes setjmp() and this triggers the -Wclobbered
+ * warning for the iRet variable.
+ */
+static void
+waitForWorkers(void)
+{
+       pthread_mutex_lock(&wrkrMut);
+       pthread_cleanup_push(mutexCancelCleanup, &wrkrMut);
+       while(wrkrRunning > 0) {
+               pthread_cond_wait(&wrkrIdle, &wrkrMut);
+       }
+       pthread_cleanup_pop(1);
+}
 
 /* Process a workset, that is handle io. We become activated
  * from either select or epoll handler. We split the workload
@@ -779,12 +793,7 @@ processWorkset(tcpsrv_t *pThis, nspoll_t *pPoll, int numEntries, nsd_epworkset_t
                 * rest of this module can not handle the concurrency introduced
                 * by workers running during the epoll call.
                 */
-               pthread_mutex_lock(&wrkrMut);
-               pthread_cleanup_push(mutexCancelCleanup, &wrkrMut);
-               while(wrkrRunning > 0) {
-                       pthread_cond_wait(&wrkrIdle, &wrkrMut);
-               }
-               pthread_cleanup_pop(1);
+               waitForWorkers();
        }
 
 finalize_it:
@@ -931,6 +940,8 @@ Run(tcpsrv_t *pThis)
         * handler in case we are cancelled.
         */
        pthread_cleanup_push(RunCancelCleanup, (void*) &pPoll);
+       /* Reset iRet to avoid warning about it being clobbered by longjmp */
+       iRet = RS_RET_OK;
 
        if((localRet = nspoll.Construct(&pPoll)) == RS_RET_OK) {
                if(pThis->pszDrvrName != NULL)
```

rgerhards commented:
CI has a false positive and has been changed quite a bit since this was introduced.

@Frogging101 thx again for the PR, finally able to merge :-)
